src/ucp/core/ucp_worker.c

```diff
@@ -3090,38 +3089,48 @@ void ucp_worker_keepalive_remove_ep(ucp_ep_h ep)
         return;
     }
 
-    round_inprogress = worker->keepalive.iter != worker->keepalive.iter_end;
+    if ((worker->keepalive.iter == &ucp_ep_ext_gen(ep)->ep_list) &&
+        (worker->keepalive.iter_end == &ucp_ep_ext_gen(ep)->ep_list)) {
+        /* Move the iterator and the stop element to the previous element and
+         * complete the current keepalive round if it is in-progress.
+         */
+        ucs_debug("worker %p: removed the keepalive current/stop endpoint %p",
+                  worker, ep);
 
-    if (worker->keepalive.iter == &ucp_ep_ext_gen(ep)->ep_list) {
-        /* Set lane_map=0 to make sure the endpoint won't be selected again */
-        ucs_debug("worker %p: removed keepalive current ep %p, moving to next",
+        worker->keepalive.iter     = worker->keepalive.iter_end->prev;
+        worker->keepalive.iter_end = worker->keepalive.iter;
+        ucs_assert(worker->keepalive.iter != &ucp_ep_ext_gen(ep)->ep_list);
+
+        if (worker->keepalive.lane_map != 0) {
+            ucs_assert((ucs_get_time() - worker->keepalive.last_round) >=
+                               worker->context->config.ext.keepalive_interval);
+            worker->keepalive.lane_map = 0;
+            ucp_worker_keepalive_complete(worker, ucs_get_time());
+        } else {
+            /* Keepalive is not in-progress */
+        }
+    } else if (worker->keepalive.iter == &ucp_ep_ext_gen(ep)->ep_list) {
+        /* Move the iterator to the next element to continue doing keeaplive
+         * starting from the next element during progressing since its lane
+         * map isn't empty (if it is empty, keepalive will just go to the next
+         * element).
+         */
+        ucs_debug("worker %p: removed the keepalive current endpoint %p",
                   worker, ep);
+
+        /* Set lane_map=0 to make sure the endpoint won't be selected again */
         worker->keepalive.lane_map = 0;
         ucp_worker_keepalive_next_ep(worker);
-        ucs_assert(worker->keepalive.iter != &ucp_ep_ext_gen(ep)->ep_list);
-    }
+    } else if (worker->keepalive.iter_end == &ucp_ep_ext_gen(ep)->ep_list) {
```

Comment:
/* Removing last but not current - move iter_end to previous ep */

Reply:
done